File: epi_judge_java/epi/PrimitiveMultiply.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class PrimitiveMultiply {
  @EpiTest(testDataFile = "primitive_multiply.tsv")
  public static long multiply(long x, long y) {
    if (x == 0 || y == 0) {
      return 0;
    }

    long sum = 0;
    while (y != 0) {
      if ((y & 1) == 1) {
        sum = primitiveAdd(sum, x);
      }
      x <<= 1;
      y >>>= 1;
    }

    return sum;
  }

  public static long primitiveAdd(long x, long y) {
    while (y != 0) {
      long carry = x & y;
      long diff = x ^ y;
      x = (carry << 1) | diff;
      y = (carry << 1) & diff;
    }

    return x;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "PrimitiveMultiply.java",
                new Object() {
                }.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_python/primitive_multiply.py
from test_framework import generic_test


def multiply(x: int, y: int) -> int:
    sum = 0
    while y != 0:
        if y & 1:
            sum = add(sum, x)
        x <<= 1
        y >>= 1
    return sum


def add(x: int, y: int) -> int:
    while y != 0:
        shift = x & y
        carry = x ^ y
        x = (shift << 1) | carry
        y = (shift << 1) & carry

    return x


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('primitive_multiply.py',
                                       'primitive_multiply.tsv', multiply))
